File: cpp/String/L_optimalPartitionString.cpp
#include<bits/stdc++.h>
using namespace std;

int partitionString(string s) {
        int idx = 0;
        int count = 0;
        unordered_map<char, bool> mp; // unordered_map to keep track of characters in current substring
        while (idx < s.length()) {
            if (mp.find(s[idx]) != mp.end()) { // if current character has already appeared in current substring
                count++; // increment count
                mp.clear(); // clear unordered_map to start a new substring
            }
            mp[s[idx]] = true; // mark current character as seen
            idx++; // move to next character
        }
        return count + 1;
    }

File: Python/Strings/L_optimalPartitionString.py
def partitionString(s: str) -> int:
        idx = 0
        count = 0
        mp = {} # use a dictionary instead of a HashMap
        while idx < len(s):
            if s[idx] in mp: # if current character has already appeared in current substring
                count += 1 # increment count
                mp.clear() # clear dictionary to start a new substring
            mp[s[idx]] = True # mark current character as seen
            idx += 1 # move to next character
        return count + 1